### PR TITLE
Refactor PDO tests with client buffer or table with all datatypes

### PR DESCRIPTION
**Pull request by v-kaywon**
This change is [Reviewable](https://reviewable.io/reviews/microsoft/msphpsql/582)

### CODE REVIEW

File: test/functional/pdo_sqlsrv/pdo_testDataType.phpt [hunk 1]
```diff
@@ -1,494 +0,0 @@
---TEST--
-Test the different type of data for retrieving
```
Comment (v-kaywon) [inline test/functional/pdo_sqlsrv/pdo_testDataType.phpt]: This test was deleted because it is a duplication of pdostatement_GetDataType.phpt

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+fwrite($bin, '00');
+rewind($bin);
+$binary_col = array($bin, $bin);
+
+$string_col = array("STRINGCOL1", "STRINGCOL2");
+
+$decimal_col = array(111.111, 222.222);
+
+$large_string = "This is a really large string used to test certain large data types like xml data type. The length of this string is greater than 256 to correctly test a large data type. This is currently used by atleast varchar type and by xml type. The fetch tests are the primary consumer of this string to validate that fetch on large types work fine. The length of this string as counted in terms of number of characters is 417.";
+
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: This is the original data in MsData.inc. For this PR, I won't change it, but maybe we can add a task for testing with more sophisticated data on the PDO side.

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+    }
+}
+
+function createTableMainTypes($conn, $tbname)
+{
+    try {
+        require_once("MsCommon_mid-refactor.inc");
+        if (isColEncrypted()) {
+            $xmlType = "nvarchar(max)";
+        } else {
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: why? xml is not supported for encrypted columns.

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+        $columnMetaArr = array("IntCol" => "int",
+                               "CharCol" => "char(10)",
+                               "NCharCol" => "nchar(10)",
+                               "DateTimeCol" => "datetime",
+                               "VarcharCol" => "varchar(50)",
+                               "NVarCharCol" => "nvarchar(50)",
+                               "FloatCol" => "float",
+                               "XmlCol" => $xmlType);
+        createTable($conn, $tbname, $columnMetaArr, "ON [PRIMARY]");
+    } catch (Exception $e) {
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: :P too bad. This is the original table schema

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+
+        $inputs = array("IntCol" => $int_col[$rownum],
+                        "CharCol" => $string_col[$rownum],
+                        "NCharCol" => $string_col[$rownum],
+                        "DateTimeCol" => $datetime_col[$rownum],
+                        "VarcharCol" => $string_col[$rownum],
+                        "NVarCharCol" => $string_col[$rownum],
+                        "FloatCol" => $decimal_col[$rownum],
+                        "XmlCol" => $xml_col[$rownum]);
+        $stmt = insertRow($conn, $tbname, $inputs, "prepareBindParam");
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: again this is originally in MsData.inc

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+            $xmlType = "nvarchar(max)";
+        } else {
+            $moneyType = "money";
+            $smallmoneyType = "smallmoney";
+            $imageType = "image";
+            $ntextType = "ntext";
+            $textType = "text";
+            $xmlType = "xml";
+        }
+
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: but these datatypes are not supported for encrypted columns

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 3 above)
[the remark this answers is not in the document]
Comment (yitam) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: yes, just thought we should test xml type even with AE enabled.

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 2 above)
[the remark this answers is not in the document]
Comment (yitam) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: Sure please. No doubt this is way too simple.

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 6 above)
[the remark this answers is not in the document]
Comment (yitam) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: Please see my reasons for xml types

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yitam) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: Why don't you overwrite MsData.inc? Why creating a new include file?

File: test/functional/pdo_sqlsrv/pdo_error.phpt [hunk 7]
```diff
@@ -1,28 +1,27 @@
...
+require_once("MsData_PDO_AllTypes.inc");
+
+try {
     $db = connect();
+    $tbname = "PDO_MainTypes";
+    createTableMainTypes($db, $tbname);
     // query with a wrong column name.
-    $db->query( "Select * from " . $table1 . " where IntColX = 1"   );
-}
+    $db->query("Select * from " . $tbname . " where IntColX = 1");
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_error.phpt]: Most tests use UPPER CASES for T-SQL keywords, so please change this as well.

File: test/functional/pdo_sqlsrv/pdo_insertDataType.phpt [hunk 8]
```diff
@@ -1,50 +1,59 @@
...
+                    "NVarCharMaxCol" => '1 This is a really large string used to test certain large data types like xml data type. The length of this string is greater than 256 to correctly test a large data type. This is currently used by atleast varchar type and by xml type. The fetch tests are the primary consumer of this string to validate that fetch on large types work fine. The length of this string as counted in terms of number of characters is 417.');
+    $stmt = insertRow($db, $tbname, $inputs, "prepareBindParam");
+
+    dropTable($db, $tbname);
+    unset($stmt);
+    unset($conn);
+    echo "Insert complete!\n";
+} catch (PDOException $e) {
     var_dump($e);
 }
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_insertDataType.phpt]: So this test only inserts and nothing else?

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 3 above)
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: how do I test it if it's not supported? no dot encrypt the column explicitly?

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 5 above)
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: because we're in the middle of refactoring. Some not yet refactored tests may rely on the original MsData.inc

File: test/functional/pdo_sqlsrv/pdo_transaction.phpt [hunk 9]
```diff
@@ -3,51 +3,69 @@ test transaction rollback and commit
...
+    if (!isColEncrypted()) {
+        $rows = $conn->exec("DELETE FROM $tbname WHERE col1 = 'a'");
+    } else {
+        // needs to find parameter for encrypted columns
+        $sql = "DELETE FROM $tbname WHERE col1 = ?";
+        $stmt = $conn->prepare($sql);
+        $col1 = "a";
+        $stmt->execute(array($col1));
+        $rows = $stmt->rowCount();
+    }
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_transaction.phpt]: This assumes the delete will always succeed? No exception handling?

File: test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt [hunk 10]
```diff
@@ -1,200 +1,206 @@
...
-	$stmt->bindParam(':value',$data);
-	$stmt->execute();
-	$result = $stmt->fetchColumn(2);	// Retrieve BitCol
-	$result = str_replace("\0","",$result);
-	print("$result\n");
-	
+    // Binding STR with buffered queries activated.
+    // fopen returns a resource, if using PDO::PARAM_STR, the PDO side converts it to a zend_string "Resource id #*"
+    // thus a resource is not meant to bind parameter using PDO::PARAM_STR
+
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt]: Does that mean the design of the original test is flawed? So even with PDO::SQLSRV_ENCODING_BINARY specified doesn't work as expected?

File: test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt [hunk 11]
```diff
@@ -1,200 +1,206 @@
...
+    $tbname = "PDO_AllTypes";
+    createAndInsertTableAllTypes($db, $tbname);
+    insert($db, $tbname);
+    bindPARAM_LOB($db, $tbname);
+    bindPARAM_STR($db, $tbname);
+    bindPARAM_NULL($db, $tbname);
+    bindPARAM_INT($db, $tbname);
+    bindPARAM_BOOL($db, $tbname);
+    delete($db, $tbname);
+
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt]: Consider not to capitalized words in the function names except LOB (an acronym). Also, please remove the underscores

File: test/functional/pdo_sqlsrv/pdostatement_GetDataType.phpt [hunk 12]
```diff
@@ -1,305 +1,302 @@
...
+function testVarcharMax($db, $tbname)
 {
-	$stmt = $db->query("SELECT XmlCol FROM PDO_AllTypes");
-	$result = $stmt->fetch(PDO::FETCH_ASSOC);
-	var_dump($result);
+    $stmt = $db->query("SELECT VarcharMaxCol FROM $tbname");
+    $result = $stmt->fetch(PDO::FETCH_ASSOC);
+    var_dump($result);
 }
-
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_GetDataType.phpt]: Please refactor this test by having only one method that also takes the column name. Most of the above methods are the same

File: test/functional/pdo_sqlsrv/pdostatement_GetDataType.phpt [hunk 13]
```diff
@@ -1,305 +1,302 @@
...
+    echo "Test_3 : int data type :\n";
+    testInt($db, $tbname);
+    echo "Test_4 : smallint data type:\n";
+    testSmallInt($db, $tbname);
+    echo "Test_5 : tinyint data type:\n";
+    testTinyInt($db, $tbname);
+    echo "Test_6 : decimal data type:\n";
+    testDecimal($db, $tbname);
+    echo "Test_7 : numeric data type:\n";
+    testNumeric($db, $tbname);
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_GetDataType.phpt]: Please use for loops and array to further simply this large block of very similar lines

File: test/functional/pdo_sqlsrv/pdostatement_bindColumn.phpt [hunk 14]
```diff
@@ -1,53 +1,50 @@
...
-    bindColumn_byNumber($db);
-    
-}
-catch (PDOException $e)
-{
+    bindColumn_byNumber($db, $tbname);
+
+    dropTable($db, $tbname);
+    unset($db);
+} catch (PDOException $e) {
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_bindColumn.phpt]: Please remove underscores in function names and use camelStyle.

File: test/functional/pdo_sqlsrv/PDO25_ErrorInfo.phpt [hunk 15]
```diff
@@ -5,112 +5,81 @@ Verification of PDO::errorInfo()
...
     $stmt1->closeCursor();
 
-//  @$conn1->query("SELECT id, label FROM [$table2]");
-//  CheckError(12, $conn1, '42S02');
-//  CheckError(13, $conn2, '42S02');
-    CheckError(14, $stmt1);
+    @$conn1->query("SELECT id, label FROM [$table2]");
+    checkError(12, $conn1, '42S02');
+    checkError(13, $conn2, '42S02');
+    checkError(14, $stmt1);
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/PDO25_ErrorInfo.phpt]: I actually don't know. Probably someone commented it out for debugging and just left it there.

File: test/functional/pdo_sqlsrv/pdostatement_closeCursor.phpt [hunk 16]
```diff
@@ -1,28 +1,27 @@
...
+    createAndInsertTableAllTypes($db, $tbname);
+    $stmt = $db->prepare("SELECT * FROM $tbname");
     $stmt->execute();
     $result = $stmt->fetch();
-    $stmt->closeCursor();
-    echo "Test complete!";
-}
-catch ( PDOException $e)
-{
+
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_closeCursor.phpt]: probably not very important, but why did you remove closeCursor()?

File: test/functional/pdo_sqlsrv/pdostatement_debugDumpParams.phpt [hunk 17]
```diff
@@ -1,26 +1,30 @@
...
+try {
     $db = connect();
-    global $table2;
+    $tbname = "PDO_AllTypes";
+    createAndInsertTableAllTypes($db, $tbname);
     $bigint = 1;
     $string = "STRINGCOL1";
-    $stmt = $db->prepare("SELECT IntCol FROM " . $table2 . " WHERE BigIntCol = :bigint AND CharCol = :string");
+    $stmt = $db->prepare("SELECT IntCol FROM " . $tbname . " WHERE BigIntCol = :bigint AND CharCol = :string");
     $stmt->bindValue(':bigint', $bigint, PDO::PARAM_INT);
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_debugDumpParams.phpt]: "SELECT IntCol FROM $tbname WHERE..."

File: test/functional/pdo_sqlsrv/pdostatement_fetchColumn.phpt [hunk 18]
```diff
@@ -1,54 +1,42 @@
...
+
+try {
+    $db = connect();
+    $tbname = "PDO_MainTypes";
+    createAndInsertTableMainTypes($db, $tbname);
+
+    $stmt = $db->query("Select * from $tbname");
+
+    // Fetch the first column from the next row in resultset. (This wud be first row since this is a first call to fetchcol)
+    $result = $stmt->fetchColumn();
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_fetchColumn.phpt]: Please use _would_, not _wud_ :)

File: test/functional/pdo_sqlsrv/pdo_insertDataType.phpt (hunk 8 above)
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdo_insertDataType.phpt]: yes, there is another test that inserts and selects

File: test/functional/pdo_sqlsrv/pdostatement_fetch_style.phpt [hunk 19]
```diff
@@ -81,73 +73,69 @@ function fetch_bound( $conn )
...
+    try {
         $result = $stmt->fetch(PDO::FETCH_UNKNOWN);
-    }
-    catch(PDOException $err)
-    {
+    } catch (PDOException $err) {
         print_r($err);
-    }   
+    }
 }
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_fetch_style.phpt]: Please refactor this test some more as many methods are very similar

File: test/functional/pdo_sqlsrv/pdostatement_getColumnMeta.phpt [hunk 20]
```diff
@@ -1,76 +1,81 @@
 --TEST--
 Test the PDOStatement::getColumnMeta() method (Note: there could be an issue about using a non-existent column index --- doesn't give any error/output/warning).
 --SKIPIF--
-<?php require('skipif.inc'); ?>
+<?php require('skipif_mid-refactor.inc'); ?>
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_getColumnMeta.phpt]: Reduce the test and put more words into the description instead

File: test/functional/pdo_sqlsrv/pdostatement_getColumnMeta_unicode_col_name.phpt [hunk 21]
```diff
@@ -1,124 +1,115 @@
 --TEST--
 Test the PDOStatement::getColumnMeta() method for Unicode column names (Note: there could be an issue about using a non-existent column index --- doesn't give any error/output/warning).
 --SKIPIF--
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_getColumnMeta_unicode_col_name.phpt]: The test name is too long. Use description instead.

File: test/functional/pdo_sqlsrv/pdostatement_setFetchMode.phpt [hunk 22]
```diff
@@ -1,48 +1,48 @@
...
-    echo "Set Fetch Mode for PDO::FETCH_OBJ \n";
+    $stmt = $db->query("SELECT * FROM $tbname");
+    echo "Set Fetch Mode for PDO::FETCH_OBJ\n";
     $stmt->setFetchMode(PDO::FETCH_OBJ);
     $result = $stmt->fetch();
     var_dump($result);
-} 
-catch ( PDOException $e)
-{
+} catch (PDOException $e) {
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdostatement_setFetchMode.phpt]: Please refactor using loops and arrays to simplify the test

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 3 above)
[the remark this answers is not in the document]
Comment (yitam) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: yes, just do not encrypt xml column or any other types that are not supported

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 5 above)
[the remark this answers is not in the document]
Comment (yitam) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: Are there many of them left not refactored? Perhaps refactor them all for this PR?

File: test/functional/pdo_sqlsrv/pdo_insertDataType.phpt (hunk 8 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_insertDataType.phpt]: are they very similar? If so just remove one and marked it as duplicate?

File: test/functional/pdo_sqlsrv/pdo_prepare.phpt [hunk 23]
```diff
@@ -3,94 +3,84 @@ Test the PDO::prepare() method.
 --ENV--
 PHPT_EXEC=true
 --SKIPIF--
-<?php require('skipif.inc'); ?>
+<?php require('skipif_mid-refactor.inc'); ?>
 --FILE--
 <?php
+require_once("MsCommon_mid-refactor.inc");
+require_once("MsData_PDO_AllTypes.inc");
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdo_prepare.phpt]: I pulled this out explicitly so in the future we know which tests use table with all datatypes. If I need to change the name of the file later I'll just do replace all :)

File: test/functional/pdo_sqlsrv/pdo_prepare.phpt [hunk 24]
```diff
@@ -3,94 +3,84 @@ Test the PDO::prepare() method.
...
+
     // Test_1 : Test with no parameters
     echo "Test_1 : Test with no parameters :\n";
-    $stmt = $db->prepare('select * from ' . $table1);
-    if(!$stmt->execute())
-    {
+    $stmt = $db->prepare("select * from $tbname");
+    if (!$stmt->execute()) {
         die("Test_1 failed.");
     }
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdo_prepare.phpt]: The whole test is wrapped by a try catch block. I think the test is OK now since it let us know which part of the test fails.

File: test/functional/pdo_sqlsrv/pdo_prepare_emulatePrepare_float.phpt [hunk 25]
```diff
@@ -1,81 +1,82 @@
...
+        $options = array(PDO::ATTR_EMULATE_PREPARES => true);
+    }
+    $stmt = $conn->prepare($query, $options);
+    $stmt->bindParam(':c3', $c3);
+    $stmt->execute();
+    $row = $stmt->fetch(PDO::FETCH_ASSOC);
+    print_r($row);
 
-$query = "SELECT * FROM [$tableName] WHERE c3_float = :c3";
+    //with emulate prepare and encoding SQLSRV_ENCODING_SYSTEM
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdo_prepare_emulatePrepare_float.phpt]: yes

File: test/functional/pdo_sqlsrv/pdo_prepare_emulatePrepare_int.phpt [hunk 26]
```diff
@@ -1,80 +1,82 @@
...
-$query = "INSERT INTO [$tableName] (name, calories) VALUES ('banana', 175)";
-$stmt = $conn->query($query);
+    // prepare with emulate prepare
+    print_r("Prepare with emulate prepare and no bindParam options:\n");
+    if (!isColEncrypted()) {
+        // emulate prepare is not supported for encrypted columns
+        $options = array(PDO::ATTR_EMULATE_PREPARES => true);
+    }
+    $stmt = $conn->prepare($query, $options);
+    $stmt->bindParam(':cal', $cal, PDO::PARAM_INT);
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdo_prepare_emulatePrepare_int.phpt]: same as which one? pdo_prepare_emulatePrepare_float.phpt? This one is different, it binds int and uses PDO::PARAM_INT.

File: test/functional/pdo_sqlsrv/pdo_prepare_emulatePrepare_unicode.phpt [hunk 27]
```diff
@@ -1,89 +1,96 @@
...
+        $conn->exec("INSERT INTO [$tableName] (name, status, age) VALUES (N'Абрам', 1, 40)");
+        $conn->exec("INSERT INTO [$tableName] (name, status, age) VALUES (N'가각', 1, 30)");
+        $query = "SELECT * FROM [$tableName] WHERE name = :name AND status = 1";
+    } else {
+        insertRow($conn, $tableName, array("name" => "Belle", "status" => 1, "age" => 34));
+        insertRow($conn, $tableName, array("name" => "Абрам", "status" => 1, "age" => 40));
+        insertRow($conn, $tableName, array("name" => "가각", "status" => 1, "age" => 30));
+        $query = "SELECT * FROM [$tableName] WHERE name = :name AND status = :status";
+    }
+
```
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdo_prepare_emulatePrepare_unicode.phpt]: It was in the original test. I don't remember why it was there in the first place.

File: test/functional/pdo_sqlsrv/pdo_transaction.phpt (hunk 9 above)
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdo_transaction.phpt]: there is exception handling in main

File: test/functional/pdo_sqlsrv/pdostatement_closeCursor.phpt (hunk 16 above)
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdostatement_closeCursor.phpt]: I think back then when we didn't know the driver well enough, we always get confused between $stmt->closeCursor(), $stmt=null and unset($stmt). I think what the test is trying to do is to clean the statement, which I did using unset($stmt).

File: test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc (hunk 5 above)
[the remark this answers is not in the document]
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/MsData_PDO_AllTypes.inc]: I'm not sure. All the tests that use the old connect() function from MsCommon.inc could use the data in MsData.inc. This PR is already large enough I don't think it's a wise decision to refactor any more tests.

File: test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt (hunk 10 above)
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt]: Yes the original test is flawed. With PDO::PARAM_STR, the PDO side already convert it to "Resource id #" when it gets to our driver. Setting the encoding the binary only changes how we pass in "Resource id #" to the ODBC driver.

File: test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt (hunk 11 above)
Comment (v-kaywon) [reply test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt]: I left it because the constant is PDO::PARAM_LOB

File: test/functional/pdo_sqlsrv/MsCommon_mid-refactor.inc [hunk 28]
```diff
@@ -161,6 +162,19 @@ class ColumnMeta
...
     }
+
+    /**
+     * @return bool if the datatype for this column is encryptable
+     */
+    public function isEncryptableType()
+    {
+        $unsupportedTypes = array("money", "smallmoney", "image", "ntext", "text", "xml", "sql_variant");
+        if (in_array(strtolower($this->dataType), $unsupportedTypes) && !$this->forceEncrypt) {
+            return false;
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/MsCommon_mid-refactor.inc]: suggestion: test if forceEncrypt first?

File: test/functional/pdo_sqlsrv/pdo_ae_insert_pdoparam_money.phpt [hunk 29]
```diff
@@ -3,84 +3,68 @@ Test for inserting and retrieving encrypted data of money types
...
-
-try
-{
-    $conn = ae_connect();
-    $conn->setAttribute( PDO::ATTR_ERRMODE, PDO::ERRMODE_SILENT );
-
-    foreach ( $dataTypes as $dataType ) {
+try {
+    $conn = connect('', array(), PDO::ERRMODE_SILENT);
+    foreach ($dataTypes as $dataType) {
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_ae_insert_pdoparam_money.phpt]: Perhaps add a comment explaining why you set ERRMODE_SILENT
because not all the AE specific tests have this error mode

File: test/functional/pdo_sqlsrv/pdo_ae_insert_pdoparam_money.phpt [hunk 30]
```diff
@@ -3,84 +3,68 @@ Test for inserting and retrieving encrypted data of money types
...
-            $conn->query( "TRUNCATE TABLE $tbname" );
+            $conn->query("TRUNCATE TABLE $tbname");
         }
-        if ( $success )
+        if ($success) {
             echo "Test successfully done.\n";
-        DropTable( $conn, $tbname );
+        }
+        DropTable($conn, $tbname);
     }
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_ae_insert_pdoparam_money.phpt]: _dropTable()_

File: test/functional/pdo_sqlsrv/pdo_ae_insert_retrieve.phpt [hunk 31]
```diff
@@ -3,73 +3,60 @@ Test for inserting encrypted data and retrieving both encrypted and decrypted da
...
-                         new columnMeta( "date", "BirthDate", null, "randomized" ));
-    create_table( $conn, $tbname, $colMetaArr );
-    
+    $colMetaArr = array( new columnMeta("int", "PatientId", "IDENTITY(1,1)"),
+                         new columnMeta("char(11)", "SSN"),
+                         new columnMeta("nvarchar(50)", "FirstName", "NULL"),
+                         new columnMeta("nvarchar(50)", "LastName", "NULL"),
+                         new columnMeta("date", "BirthDate", null, "randomized"));
+    createTable($conn, $tbname, $colMetaArr);
+
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_ae_insert_retrieve.phpt]: _ColumnMeta_

File: test/functional/pdo_sqlsrv/pdo_ae_insert_retrieve_nvarchar.phpt [hunk 32]
```diff
@@ -1,79 +1,60 @@
...
-            if ( $decrypted_row[ 'CharCount' ] == strlen( $decrypted_row[ get_default_colname( "nvarchar(1000)" ) ] )) 
-            {
+if (isColEncrypted()) {
+    try {
+        $conn1 = connect('', array(), PDO::ERRMODE_EXCEPTION, true);
+        $stmt = $conn1->query($selectSql);
+        while ($decrypted_row = $stmt->fetch(PDO::FETCH_ASSOC)) {
+            if ($decrypted_row[ 'CharCount' ] == strlen($decrypted_row[getDefaultColName("nvarchar(1000)")])) {
                 $rowInd = $decrypted_row[ 'CharCount' ] + 1;
                 echo "Failed to encrypted at row $rowInd\n";
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_ae_insert_retrieve_nvarchar.phpt]: please remove spaces within square brackets

File: test/functional/pdo_sqlsrv/pdo_ae_insert_string.phpt [hunk 33]
```diff
@@ -3,44 +3,37 @@ Test for inserting and retrieving encrypted data of string types
...
+
         // create table
-        $tbname = GetTempTableName( "", false );
-        $colMetaArr = array( new columnMeta( $dataType, "c_det" ), new columnMeta( $dataType, "c_rand", null, "randomized" ));
-        create_table( $conn, $tbname, $colMetaArr );
-        
+        $tbname = getTableName();
+        $colMetaArr = array( new columnMeta($dataType, "c_det"), new columnMeta($dataType, "c_rand", null, "randomized"));
+        createTable($conn, $tbname, $colMetaArr);
+
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/pdo_ae_insert_string.phpt]: _ColumnMeta_

File: test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt (hunk 11 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdostatement_Buffqry.phpt]: Perhaps only remove the underscores (php coding standard)?

**Comment by v-kaywon** [comment]
@yitam , these ae tests were never tested on SQL Server 2K12 before, and seems to be more strict on decimal types. As for pdo_ae_insert_pdoparam_numeric.phpt, the initial expected values were wrong.

File: test/functional/pdo_sqlsrv/pdo_ae_insert_pdoparam_money.phpt (hunk 29 above)
Comment (yitam) [reply test/functional/pdo_sqlsrv/pdo_ae_insert_pdoparam_money.phpt]: The comment sounds unclear. Please rephrase when you have a chance.

File: test/functional/pdo_sqlsrv/MsCommon.inc [hunk 34]
```diff
@@ -447,155 +444,6 @@ function is_col_enc()
...
-  catch(Exception $e)
-  {
-     var_dump( $e);
-     exit;
-  }
-}
-
 function teardown()
 {
     // TBD
```
Comment (yitam) [inline test/functional/pdo_sqlsrv/MsCommon.inc]: There are quite a few methods in MsCommon.inc that are probably not very useful. Next time please consider merging your changes in mid-refactor to this one and removing the unused helper methods.
